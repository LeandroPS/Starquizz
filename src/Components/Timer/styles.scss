@import '../../styles/colors';

.timer{
    padding: 5px 10px;
    background-color: $light-gray;
    border-radius: 5px;
    width: 50px;
    text-align: center;
}