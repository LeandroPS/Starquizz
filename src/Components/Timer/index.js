import React, {Component} from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

class Timer extends Component{
    constructor(){
        super();

        this.state = {
            currentTime: null
        }
    }

    componentDidMount(){
        this.setState({currentTime: this.props.initial});

        this.countDown = setInterval(this.onDecrease, 1000);
    }

    componentWillUnmount(){
        clearInterval(this.countDown);
    }

    onDecrease = ()=> {
        let {currentTime} = this.state;

        if(currentTime>1){
            currentTime--;
            this.setState({currentTime});
        }else{
            this.onTimeUp();
        }
    }

    onTimeUp = () => {
        clearInterval(this.countDown);
        this.props.onTimeUp();
    }

    formatTime = (totalSeconds) =>{
        const minutes = Math.floor(totalSeconds / 60);
        const seconds = totalSeconds - minutes * 60;

        return `${("0" + minutes).slice(-2)}:${("0" + seconds).slice(-2)}`;
    }

    render(){
        const {currentTime} = this.state;

        return(
            <div className="timer">
                {this.formatTime(currentTime)}
            </div>
        );
    }
}

Timer.propTypes = {
    initial: PropTypes.number.isRequired
}

export default Timer;