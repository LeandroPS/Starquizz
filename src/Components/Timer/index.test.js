import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from "enzyme";
import { create } from "react-test-renderer";
import Timer from './';

describe('Timer', () => {
    const props = {
        initial: 10,
        onTimeUp: jest.fn()
    }

    jest.useFakeTimers();

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Timer {...props}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = create(<Timer {...props}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls onTimeUp()', () => {
        const component = shallow(<Timer {...props}/>);
        expect(props.onTimeUp).not.toHaveBeenCalled();
        jest.runAllTimers();
        expect(props.onTimeUp).toHaveBeenCalled();
    });
});