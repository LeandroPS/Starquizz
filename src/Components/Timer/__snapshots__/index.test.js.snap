// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer renders correctly 1`] = `
<div
  className="timer"
>
  00:10
</div>
`;
